src/app/qgsdockablewidgethelper.cpp

```diff
@@ -0,0 +1,244 @@
+
+#include "qgsdockablewidgethelper.h"
+
+#include "qgisapp.h"
+#include "qgsdockwidget.h"
+
+#include <QWidget>
+
+QgsDockableWidgetHelper::QgsDockableWidgetHelper( bool isDocked, QWidget *widget )
+  : QWidget( nullptr )
+  , mWidget( widget )
+  , mDialogGeometry( 0, 0, 200, 200 )
+  , mDockGeometry( QRect() )
+  , mIsDockFloating( true )
+  , mDockArea( Qt::RightDockWidgetArea )
+{
+  toggleDockMode( isDocked );
+}
+
+QgsDockableWidgetHelper::~QgsDockableWidgetHelper()
+{
+  if ( mDock )
+  {
+    mDockGeometry = mDock->geometry();
+    mIsDockFloating = mDock->isFloating();
+    mDockArea = QgisApp::instance()->dockWidgetArea( mDock );
+
+    mDock->setWidget( nullptr );
+    QgisApp::instance()->removeDockWidget( mDock );
+    delete mDock;
+    mDock = nullptr;
+  }
+
+  if ( mDialog )
+  {
+    mDialogGeometry = mDialog->geometry();
+
+    mDialog->layout()->removeWidget( mWidget );
+    delete mDialog;
+    mDialog = nullptr;
+  }
+}
+
+void QgsDockableWidgetHelper::setWidget( QWidget *widget )
+{
+  // Make sure the old mWidget is not stuck as a child of mDialog or mDock
+  if ( mWidget )
+  {
+    mWidget->setParent( QgisApp::instance() );
+  }
+  if ( mDialog )
+  {
+    mDialog->layout()->removeWidget( mWidget );
+  }
+  if ( mDock )
+  {
+    mDock->setWidget( nullptr );
+  }
+
+  mWidget = widget;
+  toggleDockMode( mIsDocked );
+}
+
+bool QgsDockableWidgetHelper::isDocked() const
+{
+  return mIsDocked;
+}
+
+void QgsDockableWidgetHelper::toggleDockMode( bool docked )
+{
+  // Make sure the old mWidget is not stuck as a child of mDialog or mDock
+  if ( mWidget )
+  {
+    mWidget->setParent( QgisApp::instance() );
+  }
+
+  // Remove both the dialog and the dock widget first
+  if ( mDock )
+  {
+    mDockGeometry = mDock->geometry();
+    mIsDockFloating = mDock->isFloating();
+    mDockArea = QgisApp::instance()->dockWidgetArea( mDock );
+
+    mDock->setWidget( nullptr );
+    QgisApp::instance()->removeDockWidget( mDock );
+    delete mDock;
+    mDock = nullptr;
+  }
+
+  if ( mDialog )
+  {
+    mDialogGeometry = mDialog->geometry();
+
+    mDialog->layout()->removeWidget( mWidget );
+    delete mDialog;
+    mDialog = nullptr;
+  }
+
+  mIsDocked = docked;
+
+  // If there is no widget set, do not create a dock or a dialog
+  if ( mWidget == nullptr )
+    return;
+
+  if ( docked )
+  {
+    // going from window -> dock
+    mDock = new QgsDockWidget( QgisApp::instance() );
+    mDock->setWindowTitle( this->windowTitle() );
+    mDock->setWidget( mWidget );
+    setupDockWidget();
+
+    connect( mDock, &QgsDockWidget::closed, [ = ]()
+    {
+      mDockGeometry = mDock->geometry();
+      mIsDockFloating = mDock->isFloating();
+      mDockArea = QgisApp::instance()->dockWidgetArea( mDock );
+      QgisApp::instance()->close3DMapView( this->windowTitle() );
+    } );
+  }
+  else
+  {
+    // going from dock -> window
+    mDialog = new QDialog( QgisApp::instance(), Qt::Window );
```

Comment:
I am not very happy with all the `QgisApp::instance()` calls in this class. How about we make it more generic - and in constructor the caller would pass a pointer to `QMainWindow` so that we don't need to use `QgisApp` here